daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/dpath/NodeInfo.scala

```diff
@@ -198,58 +199,39 @@ object NodeInfo extends Enum {
     }
 
     def isError: Boolean = false
-    def primType = this
+
+    private lazy val optPrimType_ = Some(this)
+    override def optPrimType: Option[PrimType] = optPrimType_
 
     def fromXMLString(s: String): DataValuePrimitive
 
     override def toString = name
   }
 
-  private def getTypeNode(name: String) = {
-    val namelc = name.toLowerCase()
-    allTypes.find(stn => stn.lcaseName == namelc)
-  }
-
-  /**
-   * For Java API use, we have a very restricted trait api.PrimitiveType
-   * mixed into PrimTypeNode, so that we can hand PrimTypeNode as result
-   * from methods callable from Java without exposing all of PrimType's
-   * implementation.
-   * @param name lookup key, case insensitive
-   * @return an api.PrimitiveType, or null if there is no type with that name.
-   */
-  def primitiveTypeFromName(name: String): PrimitiveType = {
-    allDFDLTypesLookupTable.get(name)
-  }
-
-  def isXDerivedFromY(nameX: String, nameY: String): Boolean = {
-    if (nameX == nameY) true
-    else {
-      getTypeNode(nameX) match {
-        case Some(stn) => {
-          stn.doesParentListContain(nameY)
-        }
-        case None => false
-      }
-    }
-  }
-
   sealed trait Kind extends EnumValueType with PrimTypeView {
 
     def name: String = Misc.getNameFromClass(this)
 
     def parents: Seq[Kind]
     def children: Seq[Kind]
 
-    final lazy val isHead: Boolean = parents.isEmpty
-    final lazy val lcaseName = name.toLowerCase()
+    /**
+     * The prim type corresponding to this type.
+     * So if this Kind is derived from a PrimType kind, then
+     * this kind should return that prim type object.
+     *
+     * @return None if there is no corresponding prim type.
+     */
+    def optPrimType: Option[PrimType] = None
 
-    final lazy val selfAndAllParents: Set[Kind] = parents.flatMap {
+    private final lazy val lcaseName = name.toLowerCase()
```

Comment:
Is there value in storing the lowercase version of this as a member? I feel like memory is our big problem, everytime we add a new member we are taking up just a little bit more memory. In fact, it's only used by `parentList`, and that is only used by `doesParentListContain`, which just calls `toLowerCase` again, so we aren't even saving any CPU by storing it. Probably a premature optimization, but seems like `doesParentListContain` could be this:

```scala
    def doesParentListContain(typeName: String): Boolean = {
      selfAndParents.exists { _.name.equalsIngoreCase(typeName) }
    }
```
Avoids both he lcaseName and parentList members, and I imagine is just as fast.

All that said, I now see `doesParentListContain` is never even used, so I suggest we just drop that function and the associated members it uses.

Reply:
I will drop this method and associated.

Note that this graph of NodeInfo TypeNode objects is basically a tiny static data structure. There are fewer than 30 such objects. Everything else just points at these objects or uses their types (called Kinds)  to represent a DFDL type or other static characteristic of data or an argument/result of an expression.

Reply:
Yeah, I don't think memory is really an issue in this case. Just something I try to be aware of. I don't have any numbers, but I feel like we have a lot of things that are calculated once and stored in a val and never used again just taking up memory.